File: TerraObserver/Godot.Abstractions/Bases/INode3D.cs
namespace Godot.Abstractions.Bases;

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH
/// Date: 2025-04-16 16:51:16
public interface INode3D : INode
{
    bool Visible { get; set; }
    void Show();
    void Hide();
    Transform3D Transform { get; set; }
    Transform3D GlobalTransform { get; set; }
    Vector3 Position { get; set; }
    Vector3 Rotation { get; set; }
    Vector3 RotationDegrees { get; set; }
    Quaternion Quaternion { get; set; }
    Basis Basis { get; set; }
    Vector3 Scale { get; set; }
    Vector3 GlobalPosition { get; set; }
    Basis GlobalBasis { get; set; }

    World3D GetWorld3D();
    Vector3 ToLocal(Vector3 globalPoint);
    Vector3 ToGlobal(Vector3 localPoint);

    void Rotate(Vector3 axis, float angle);
    void GlobalRotate(Vector3 axis, float angle);
    void RotateX(float angle);
}

File: TerraObserver/TO.Domains/Types/Godots/INode3D.fs
namespace TO.Domains.Types.Godots

open System
open System.Runtime.InteropServices
open Godot

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH
/// Date: 2025-04-16 16:51:16
[<Interface>]
type INode3D =
    inherit INode
    abstract Transform: Transform3D with get, set // 59
    abstract GlobalTransform: Transform3D with get, set // 74
    abstract Position: Vector3 with get, set // 89
    abstract Rotation: Vector3 with get, set // 106
    abstract RotationDegrees: Vector3 with get, set // 121
    abstract Quaternion: Quaternion with get, set // 136
    abstract Basis: Basis with get, set // 151
    abstract Scale: Vector3 with get, set // 168
    abstract GlobalPosition: Vector3 with get, set // 228
    abstract GlobalBasis: Basis with get, set // 243
    abstract Visible: bool with get, set // 289
    abstract GetWorld3D: unit -> World3D // 676
    abstract Show: unit -> unit // 815
    abstract Hide: unit -> unit // 826
    abstract Rotate: axis: Vector3 * angle: float32 -> unit // 881
    abstract GlobalRotate: axis: Vector3 * angle: float32 -> unit // 892
    abstract RotateX: angle: float32 -> unit // 958

    abstract LookAt:
        target: Vector3 *
        [<Optional; DefaultParameterValue(Nullable<Vector3>())>] up: Vector3 Nullable *
        [<Optional; DefaultParameterValue(false)>] useModelFront: bool ->
            unit // 1031

    abstract ToLocal: globalPoint: Vector3 -> Vector3 // 1056
    abstract ToGlobal: localPoint: Vector3 -> Vector3 // 1067
